include/buffer.hpp

```diff
@@ -0,0 +1,117 @@
+#ifndef GNURADIO_BUFFER2_H
+#define GNURADIO_BUFFER2_H
+
+#include <bit>
+#include <cstdint>
+#include <type_traits>
+#include <concepts>
+#include <span>
+
+namespace gr {
+namespace util { // TODO: move this meta-programming helpers to a common header file
+template <typename T, typename...>
+struct first_template_arg_helper;
+
+template <template <typename...> class TemplateType,
+          typename ValueType,
+          typename... OtherTypes>
+struct first_template_arg_helper<TemplateType<ValueType, OtherTypes...>> {
+    using value_type = ValueType;
+};
+
+template <typename T>
+constexpr auto* value_type_helper()
+{
+    if constexpr (requires { typename T::value_type; }) {
+        return static_cast<typename T::value_type*>(nullptr);
+    }
+    else {
+        return static_cast<typename first_template_arg_helper<T>::value_type*>(nullptr);
+    }
+}
+
+template <typename T>
+using value_type_t = std::remove_pointer_t<decltype(value_type_helper<T>())>;
+
+template <typename... A>
+struct test_fallback {
+};
+
+template <typename, typename ValueType>
+struct test_value_type {
+    using value_type = ValueType;
+};
+
+static_assert(std::is_same_v<int, value_type_t<test_fallback<int, float, double>>>);
+static_assert(std::is_same_v<float, value_type_t<test_value_type<int, float>>>);
+static_assert(std::is_same_v<int, value_type_t<std::span<int>>>);
+static_assert(std::is_same_v<double, value_type_t<std::array<double, 42>>>);
+
+constexpr bool isPower2(std::unsigned_integral auto value)
+{
+    return !(value == 0) && !(value & (value - 1));
+}
+template <std::unsigned_integral auto N>
+static constexpr bool is_power2_v = isPower2(N);
```

Comment:
You could use std::has_single_bit(N) from `<bit>`.

Reply:
Tried this because I saw this also on your bits of the code but didn't evaluate in that context... will re-check..

Reply:
missed a bracket in my earlier attempt... done.